File: design_patterns/half_sync_half_async/half_sync_half_async_python/src/half_sync_half_async_simplified_demo.py
import threading, time
from typing import Optional
from queue import Queue, Empty
from concurrent.futures import Future, ThreadPoolExecutor


class Task:

    def __init__(self):
        self.result = None

    def execute(self):
        time.sleep(1)
        thread = threading.current_thread()
        self.result = "Task %s is executing in thread %s" % (self, thread.name)
        return self

    def on_complete(self):
        print(self.result)
        thread = threading.current_thread()
        print("Executing completion handler in thread %s for task %s" % (thread.name, self))


class AsynchronousService:

    def __init__(self, synchronous_service: ThreadPoolExecutor):
        self.completed_task_queue = Queue()
        self.synchronous_service = synchronous_service

    def submit(self, task: Task):
        future: Future = self.synchronous_service.submit(task.execute)

        def add_to_completion_queue(completed_future: Future):
            self.completed_task_queue.put(completed_future.result())

        future.add_done_callback(add_to_completion_queue)

    def get_completed_task(self) -> Optional[Task]:
        try:
            return self.completed_task_queue.get_nowait()
        except Empty:
            return None


# This is a Python-based pseudocode written with the intent
# to provide intuition about the Half-sync/half-async pattern
if __name__ == '__main__':
    synchronous_service = ThreadPoolExecutor(max_workers=10)
    asynchronous_service = AsynchronousService(synchronous_service)

    asynchronous_service.submit(Task())

    # There should be no completed tasks at this point, since the only previously
    # submitted task sleeps for 1000ms
    completed_task = asynchronous_service.get_completed_task()
    assert completed_task is None

    asynchronous_service.submit(Task())

    expected_tasks_to_complete = 2
    while expected_tasks_to_complete > 0:
        completed_task = asynchronous_service.get_completed_task()
        if completed_task is not None:
            completed_task.on_complete()
            expected_tasks_to_complete -= 1
        # This sleep here is just for demo purposes, in real-life situation
        # it would be some other logic executed by the main thread until
        # there are completed tasks in the queue, which it can pick up and run
        # completion handlers
        time.sleep(1)
        # ... do something else

    synchronous_service.shutdown()

File: design_patterns/half_sync_half_async/half_sync_half_async_java/src/main/java/app/swarch/designpatterns/halfsynchalfasync/simplified/HalfSyncHalfAsyncSimplifiedDemo.java
package app.swarch.designpatterns.halfsynchalfasync.simplified;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

abstract class Task<R> {

    protected R result;

    abstract R executeSynchronously();

    abstract void onComplete();

    public void saveResult(R result) {
        this.result = result;
    }
}

class AsynchronousService {

    private final ThreadPoolExecutor synchronousService;
    private final BlockingQueue<Task<?>> completedTaskQueue = new LinkedBlockingQueue<>();

    public AsynchronousService(ThreadPoolExecutor synchronousService) {
        this.synchronousService = synchronousService;
    }

    public <R> void submit(Task<R> task) {
        synchronousService.execute(() -> {
            try {
                R result = task.executeSynchronously();
                task.saveResult(result);
                completedTaskQueue.put(task);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
    }

    public Task<?> pollCompletedTask() {
        return completedTaskQueue.poll();
    }
}

public class HalfSyncHalfAsyncSimplifiedDemo {

    public static void main(String[] args) throws Exception {
        BlockingQueue<Runnable> taskQueue = new LinkedBlockingQueue<>();
        ThreadPoolExecutor synchronousService = new ThreadPoolExecutor(10, 10, 10, TimeUnit.SECONDS, taskQueue);
        AsynchronousService asynchronousService = new AsynchronousService(synchronousService);

        class DemoTask extends Task<String> {
            @Override
            String executeSynchronously() {
                try {
                    Thread.sleep(1000);
                    return String.format("Executed task %s synchronously in the thread '%s'", this.hashCode(), Thread.currentThread().getName());
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return null;
            }

            @Override
            void onComplete() {
                System.out.println(this.result);
                System.out.printf("Completed task %s in the thread '%s'%n", this.hashCode(), Thread.currentThread().getName());
            }
        }

        // Submit the first task for processing
        asynchronousService.submit(new DemoTask());

        // There should be no completed tasks at this point, since the only previously
        // submitted task sleeps for 1000ms
        Task<?> completedTask = asynchronousService.pollCompletedTask();
        assert completedTask == null;

        // Submit the second task for processing
        asynchronousService.submit(new DemoTask());

        int expectedTasksToComplete = 2;
        while (expectedTasksToComplete > 0) {
            completedTask = asynchronousService.pollCompletedTask();
            if (completedTask != null) {
                completedTask.onComplete();
                expectedTasksToComplete--;
            }
            // This sleep here is just for demo purposes, in real-life situation
            // it would be some other logic executed by the main thread until
            // there are completed tasks in the queue, which it can pick up and run
            // completion handlers
            Thread.sleep(100);
            // ... do something else
        }

        synchronousService.shutdown();
    }

}
